all/core/state.py

```diff
@@ -387,6 +387,49 @@ def shape(self):
     def __len__(self):
         return self.shape[0]
 
+    @classmethod
+    def cat(cls, state_array_list, axis=0):
+        '''Concatenates along batch dimention'''
+        if len(state_array_list) == 0:
+            raise ValueError("cat accepts a non-zero size list of StateArrays")
+
+        d = {}
+        state_size = sum(state_array.shape[axis] for state_array in state_array_list)
+        new_shape = list(state_array_list[0].shape)
+        new_shape[axis] = state_size
+        new_shape = tuple(new_shape)
+        keys = list(state_array_list[0].keys())
+        for key in keys:
+            d[key] = torch.cat([state_array[key] for state_array in state_array_list], axis=axis)
+        return StateArray(d, new_shape, device=state_array_list[0].device)
+
+    def flatten_for_execution(self, fn):
+        flat_size = np.prod(self.shape)
+        flat_data = self.view((flat_size,))
+        out = fn(flat_data)
+        assert out.shape[0] == flat_size, "first dimension out output of flatten_for_execution expected to match input"
+        spread = out.view(self.shape + out.shape[1:])
+        return spread
+
+    def batch_execute(self, minibatch_size, fn):
```

Comment:
I'm wondering if `minibatch_size` should optionally accept `-1` to indicate no mini-batching, or something to that effect. Or I think it can handle infinity? Is that usage clear?

Reply:
I think you can put in an arbitrarily large number. Probably a float infinity would work. So I think it is fine as is?